File: src/loom/storage.py
"""Atomic markdown I/O with order-preserving YAML frontmatter.

The on-disk format is::

    ---
    <yaml frontmatter>
    ---
    <markdown body>

Files always end with exactly one trailing newline. Writes go via a temp
file in the destination directory followed by :func:`os.replace`, so a
``Ctrl-C`` mid-write can never leave a partially written file in place.
"""

from __future__ import annotations

import io
import os
import tempfile
from pathlib import Path
from typing import Any

from ruamel.yaml import YAML
from ruamel.yaml.error import YAMLError

from .errors import ValidationError

FENCE = "---"

_yaml = YAML(typ="rt")
_yaml.preserve_quotes = True
_yaml.default_flow_style = False


def load(path: Path) -> tuple[dict[str, Any], str]:
    """Read a markdown file, returning ``(frontmatter, body)``.

    Files without an opening ``---`` fence have an empty frontmatter and
    the entire content as body. A file that opens a fence but never
    closes one raises :class:`ValidationError`.
    """
    text = path.read_text(encoding="utf-8")
    return parse(text, source=str(path))


def dump(path: Path, frontmatter: dict[str, Any], body: str) -> None:
    """Atomically write ``frontmatter`` + ``body`` to *path*."""
    _atomic_write(path, render(frontmatter, body))


def parse(text: str, source: str | None = None) -> tuple[dict[str, Any], str]:
    """Parse a markdown string into ``(frontmatter, body)``.

    The body starts at the line *after* the closing fence; the closing
    fence is the first ``---`` line found after the opening one (so a
    body containing a horizontal rule does not confuse the parser).
    """
    lines = text.splitlines(keepends=True)

    if not lines or lines[0].rstrip("\r\n") != FENCE:
        return {}, text

    close_idx: int | None = None
    for i in range(1, len(lines)):
        if lines[i].rstrip("\r\n") == FENCE:
            close_idx = i
            break

    if close_idx is None:
        suffix = f" (source: {source})" if source else ""
        raise ValidationError(
            None,
            f"frontmatter opening '---' has no closing fence{suffix}",
        )

    yaml_text = "".join(lines[1:close_idx])
    body = "".join(lines[close_idx + 1 :])

    try:
        loaded = _yaml.load(yaml_text) if yaml_text.strip() else None
    except YAMLError as e:
        suffix = f" (source: {source})" if source else ""
        raise ValidationError(None, f"frontmatter is not valid YAML{suffix}: {e}") from e
    if loaded is None:
        loaded = {}
    if not isinstance(loaded, dict):
        raise ValidationError(
            None,
            f"frontmatter must be a YAML mapping, got {type(loaded).__name__}",
        )

    return loaded, body


def render(frontmatter: dict[str, Any], body: str) -> str:
    """Serialize ``frontmatter`` + ``body`` to the canonical on-disk form."""
    parts: list[str] = []

    if frontmatter:
        buf = io.StringIO()
        _yaml.dump(frontmatter, buf)
        yaml_text = buf.getvalue()
        if not yaml_text.endswith("\n"):
            yaml_text += "\n"
        parts.append(FENCE + "\n")
        parts.append(yaml_text)
        parts.append(FENCE + "\n")

    if body:
        parts.append(body.rstrip("\n") + "\n")

    out = "".join(parts)
    if not out.endswith("\n"):
        out += "\n"
    return out


def _atomic_write(path: Path, content: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, tmp_str = tempfile.mkstemp(prefix=f".{path.name}.", suffix=".tmp", dir=str(path.parent))
    tmp = Path(tmp_str)
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as f:
            f.write(content)
        os.replace(tmp, path)
    except BaseException:
        tmp.unlink(missing_ok=True)
        raise

File: web/lib/storage.ts
/** Atomic markdown I/O with order-preserving YAML frontmatter.
 *
 * The on-disk format is:
 *   ---
 *   <yaml frontmatter>
 *   ---
 *   <markdown body>
 *
 * Files always end with exactly one trailing newline. Writes go via a temp
 * file followed by rename, so a crash mid-write cannot leave a partial file.
 *
 * Mirrors src/loom/storage.py.
 */

import * as fs from "fs";
import * as path from "path";
import yaml from "js-yaml";
import { ValidationError } from "./errors";

export const FENCE = "---";

// ---------------------------------------------------------------------------
// parse
// ---------------------------------------------------------------------------

/**
 * Parse a markdown string into [frontmatter, body].
 *
 * Files without an opening `---` fence have empty frontmatter and the entire
 * content as body. A file that opens a fence but never closes one raises
 * ValidationError.
 */
export function parse(
  text: string,
  source?: string
): [Record<string, unknown>, string] {
  const rawLines = splitKeepingEnds(text);

  if (rawLines.length === 0 || rawLines[0]!.trimEnd() !== FENCE) {
    return [{}, text];
  }

  let closeIdx: number | null = null;
  for (let i = 1; i < rawLines.length; i++) {
    if (rawLines[i]!.trimEnd() === FENCE) {
      closeIdx = i;
      break;
    }
  }

  if (closeIdx === null) {
    const suffix = source ? ` (source: ${source})` : "";
    throw new ValidationError(
      null,
      `frontmatter opening '---' has no closing fence${suffix}`
    );
  }

  const yamlText = rawLines.slice(1, closeIdx).join("");
  const body = rawLines.slice(closeIdx + 1).join("");

  let loaded: unknown;
  try {
    loaded = yamlText.trim() ? yaml.load(yamlText) : null;
  } catch (e) {
    const suffix = source ? ` (source: ${source})` : "";
    throw new ValidationError(
      null,
      `frontmatter is not valid YAML${suffix}: ${String(e)}`
    );
  }

  if (loaded === null || loaded === undefined) {
    loaded = {};
  }
  if (typeof loaded !== "object" || Array.isArray(loaded)) {
    throw new ValidationError(
      null,
      `frontmatter must be a YAML mapping, got ${typeof loaded}`
    );
  }

  return [loaded as Record<string, unknown>, body];
}

// ---------------------------------------------------------------------------
// render
// ---------------------------------------------------------------------------

/**
 * Serialize frontmatter + body to the canonical on-disk form.
 * Key insertion order is preserved. Output ends with exactly one newline.
 */
export function render(
  frontmatter: Record<string, unknown>,
  body: string
): string {
  const parts: string[] = [];

  if (Object.keys(frontmatter).length > 0) {
    let yamlText = yaml.dump(frontmatter, {
      lineWidth: -1,
      quotingType: '"',
      forceQuotes: false,
      noRefs: true,
    });
    if (!yamlText.endsWith("\n")) yamlText += "\n";
    parts.push(FENCE + "\n");
    parts.push(yamlText);
    parts.push(FENCE + "\n");
  }

  if (body) {
    parts.push(body.replace(/\n+$/, "") + "\n");
  }

  let out = parts.join("");
  if (!out.endsWith("\n")) out += "\n";
  return out;
}

// ---------------------------------------------------------------------------
// atomic I/O
// ---------------------------------------------------------------------------

/**
 * Read a markdown file, returning [frontmatter, body].
 */
export function load(filePath: string): [Record<string, unknown>, string] {
  const text = fs.readFileSync(filePath, "utf-8");
  return parse(text, filePath);
}

/**
 * Atomically write frontmatter + body to filePath (tempfile + rename).
 */
export function dump(
  filePath: string,
  frontmatter: Record<string, unknown>,
  body: string
): void {
  atomicWriteText(filePath, render(frontmatter, body));
}

/**
 * Atomically write content to filePath via a temp file + rename.
 */
export function atomicWriteText(filePath: string, content: string): void {
  const dir = path.dirname(filePath);
  fs.mkdirSync(dir, { recursive: true });
  const base = path.basename(filePath);
  const tmpPath = path.join(dir, `.${base}.${Date.now()}.tmp`);
  try {
    fs.writeFileSync(tmpPath, content, "utf-8");
    fs.renameSync(tmpPath, filePath);
  } catch (e) {
    try {
      fs.unlinkSync(tmpPath);
    } catch {
      // ignore cleanup errors
    }
    throw e;
  }
}

// ---------------------------------------------------------------------------
// body hashing
// ---------------------------------------------------------------------------

/**
 * Return the sha256 hex digest of the raw on-disk bytes of filePath.
 * Mirrors the Python body_hash contract: hash of raw bytes, never re-rendered.
 */
export async function bodyHash(filePath: string): Promise<string> {
  const bytes = fs.readFileSync(filePath);
  const hashBuf = await crypto.subtle.digest("SHA-256", bytes);
  return Array.from(new Uint8Array(hashBuf))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Split text into lines keeping line endings attached. */
function splitKeepingEnds(text: string): string[] {
  if (text === "") return [];
  const result: string[] = [];
  let i = 0;
  while (i < text.length) {
    const nlIdx = text.indexOf("\n", i);
    if (nlIdx === -1) {
      result.push(text.slice(i));
      break;
    }
    result.push(text.slice(i, nlIdx + 1));
    i = nlIdx + 1;
  }
  return result;
}
